File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.DotNet/director/Chunks/RaysCastListChunk.cs
using Microsoft.Extensions.Logging;
using ProjectorRays.Common;
using ProjectorRays.Director;
using System.Text;

namespace ProjectorRays.director.Chunks;

public class RaysCastListChunk : RaysListChunk
{
    public ushort Unk0;
    public ushort CastCount;
    public ushort ItemsPerCast;
    public ushort Unk1;
    public List<CastListEntry> Entries = new();

    public RaysCastListChunk(RaysDirectorFile? dir) : base(dir, ChunkType.CastListChunk) { }

    public override void ReadHeader(ReadStream stream)
    {
        var check = stream.Endianness; // =  Dir?.Endianness ?? Endianness.BigEndian;

        DataOffset = stream.ReadUint32();
        //Dir?.Logger.LogInformation($"CastListChunk: DataOffset = {DataOffset}");
        Unk0 = stream.ReadUint16();
        CastCount = stream.ReadUint16();
        ItemsPerCast = stream.ReadUint16();
        Unk1 = stream.ReadUint16();
        OffsetTableLen = stream.ReadUint16();
        //Dir?.Logger.LogInformation($"CastListChunk: OffsetTableLen = {OffsetTableLen}");
        ItemsLen = stream.ReadUint32();
        //Dir?.Logger.LogInformation($"CastListChunk: ItemsLen = {ItemsLen}");
        //ItemEndianness = (Endianness)stream.ReadUint8(); <!- not part of the header

        Dir?.Logger.LogInformation($"CastListChunk: DataOffset={DataOffset}, CastCount={CastCount}, ItemsPerCast={ItemsPerCast}, OffsetTableLen={OffsetTableLen},ItemsLen={ItemsLen}, ItemEndianness={ItemEndianness}");

    }


    public override void Read(ReadStream stream)
    {
        // Use the same byte order as the movie itself. Older Director versions
        // store this table in little endian on Windows, so rely on the parent
        // file to specify the correct endianness.
        base.Read(stream);

        for (int i = 0; i < CastCount; i++)
        {
            if (i >= Items.Count || Items[i].Size == 0)
            {
                Dir?.Logger.LogWarning($"Skipping cast entry {i}: no data available.");
                Entries.Add(new CastListEntry());
                continue;
            }

            var entry = new CastListEntry();

            if (ItemsPerCast >= 1)
            {
                entry.Name = ReadPascalString(i * ItemsPerCast + 0);
            }

            if (ItemsPerCast >= 2 && (i * ItemsPerCast + 1) < Items.Count && Items[i * ItemsPerCast + 1].Size > 0)
            {
                entry.FilePath = ReadPascalString(i * ItemsPerCast + 1);
            }


            if (ItemsPerCast >= 3)
            {
                var preloadStream = new ReadStream(Items[i * ItemsPerCast + 2], ItemEndianness);
                entry.PreloadSettings = preloadStream.ReadUint16();
            }

            if (ItemsPerCast >= 4)
            {
                var extraStream = new ReadStream(Items[i * ItemsPerCast + 3], ItemEndianness);
                entry.MinMember = extraStream.ReadUint16();
                entry.MaxMember = extraStream.ReadUint16();
                entry.Id = extraStream.ReadInt32();
            }

            Entries.Add(entry);
        }
    }
    private string ReadAsciiString(BufferView buffer)
    {
        return Encoding.ASCII.GetString(buffer.Data, buffer.Offset, buffer.Size).TrimEnd('\0');
    }


    /// <summary>
    /// Serialize this cast list chunk. Each entry is written through
    /// <see cref="CastListEntry.WriteJSON"/> so we emit an array manually.
    /// </summary>
    public override void WriteJSON(RaysJSONWriter json)
    {
        json.StartObject();
        json.WriteKey("entries");
        json.StartArray();
        foreach (var e in Entries)
            e.WriteJSON(json);
        json.EndArray();
        json.EndObject();
    }
}

File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays_Python/director/chunks/rays_cast_list_chunk.py
from .rays_list_chunk import RaysListChunk
from .rays_chunk import ChunkType
from ...common.stream import ReadStream, Endianness
from ..rays_subchunk import CastListEntry

class RaysCastListChunk(RaysListChunk):
    def __init__(self, dir=None):
        super().__init__(dir, ChunkType.CastListChunk)
        self.cast_count = 0
        self.items_per_cast = 0
        self.entries = []

    def read_header(self, stream: ReadStream):
        # Cast list headers are always stored in big endian like the C# reader
        stream.set_endianness(Endianness.BigEndian)
        self.data_offset = stream.read_uint32()
        stream.read_uint16()
        self.cast_count = stream.read_uint16()
        self.items_per_cast = stream.read_uint16()
        stream.read_uint16()
        self.offset_table_len = stream.read_uint16()
        self.items_len = stream.read_uint32()
        if self.dir:
            self.dir._logger.info(
                "CastListChunk: DataOffset=%d, CastCount=%d, ItemsPerCast=%d, OffsetTableLen=%d,ItemsLen=%d, ItemEndianness=%s",
                self.data_offset,
                self.cast_count,
                self.items_per_cast,
                self.offset_table_len,
                self.items_len,
                self.item_endianness.name,
            )

    def read(self, stream: ReadStream):
        super().read(stream)
        for _ in range(self.cast_count):
            self.entries.append(CastListEntry())

    def write_json(self, writer):
        writer.start_object()
        writer.write_key('count'); writer.write_val(len(self.entries))
        writer.end_object()
